cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,259 @@
+        ax: Optional matplotlib axis to use for drawing.
+        cartesian: If True, directly position as (row, column); otherwise,
+            rotate 45 degrees to accommodate google-style diagonal grids.
+        kwargs: Additional arguments to pass to `nx.draw_networkx`.
+
+    Returns:
+        A positions dictionary mapping nodes to (x, y) coordinates suitable for future calls
+        to NetworkX plotting functionality.
+    """
+    if ax is None:
+        ax = plt.gca()  # coverage: ignore
+
+    if cartesian:
+        pos = {node: (y, -x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+    else:
+        pos = {node: (x + y, y - x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+
+    nx.draw_networkx(graph, pos=pos, ax=ax, **kwargs)
+    ax.axis('equal')
+    return pos
+
+
+@dataclass(frozen=True)
+class LineTopology(NamedTopology):
+    """A 1D linear topology.
+
+    Node indices are contiguous integers starting from 0 with edges between
+    adjacent integers.
+
+    Args:
+        n_nodes: The number of nodes in a line.
+    """
+
+    n_nodes: int
+
+    # https://github.com/python/mypy/issues/1362
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        """The name of this topology: {n_nodes}-line"""
+        return f'{self.n_nodes}-line'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        """The graph of this topology.
+
+        Node indices are contiguous integers starting from 0 with edges between
+        adjacent integers.
+        """
+        return nx.from_edgelist(
+            [(i1, i2) for i1, i2 in zip(range(self.n_nodes), range(1, self.n_nodes))]
+        )
+
+    def draw(self, ax=None, cartesian: bool = True, **kwargs) -> Dict[Any, Tuple[int, int]]:
+        """Draw this graph.
+
+        Args:
+            ax: Optional matplotlib axis to use for drawing.
+            cartesian: If True, draw as a horizontal line. Otherwise, draw on a diagonal.
+            kwargs: Additional arguments to pass to `nx.draw_networkx`.
+        """
+        g2 = nx.relabel_nodes(self.graph, {n: (n, 1) for n in self.graph.nodes})
+        return draw_gridlike(g2, ax=ax, cartesian=cartesian, **kwargs)
+
+    def _json_dict_(self):
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+
+@dataclass(frozen=True)
+class DiagonalRectangleTopology(NamedTopology):
+    width: int
+    height: int
+
+    def __post_init__(self):
+        if self.width <= 0:
+            raise ValueError("Width must be a positive integer")
+        if self.height <= 0:
+            raise ValueError("Height must be a positive integer")
+
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        return f'{self.width}-{self.height}-diagonal-rectangle'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        g = nx.Graph()
+        # construct a "diagonal rectangle graph" whose width and height
```

Comment:
Could this be a docstring instead of line comments?

Reply:
I added a greatly expanded version of this to the class docstring. Left the line comment as well with an additional note connecting it to why we use the "mega[row/col]" variable name.